varcode/coding_effect.py

```diff
@@ -162,37 +181,50 @@ def infer_coding_effect(
             ref,
             alt)
 
-    # In case sequence isn't a multiple of 3, then truncate it
-    truncated_variant_cds_len = int(len(variant_cds_seq) / 3) * 3
-    truncated_variant_cds_seq = variant_cds_seq[:truncated_variant_cds_len]
+    variant_protein = translate(variant_cds_seq)
 
-    # Can't be sure that the variant is a complete CDS, so passing cds=False
-    # in the case of a frameshift, the transcript might not actually contain
-    # a stop codon. So, we're for it manually (by passing to_stop=False).
-    variant_protein = str(Seq(truncated_variant_cds_seq).translate(
-        to_stop=False, cds=False))
+    if len(variant_protein) == 0:
+        raise ValueError(
+            "Translated mutant protein sequence of %s is empty" % (transcript,))
 
     assert len(variant_protein) > 0, \
         "Protein sequence empty for variant %s on transcript %s" % (
             variant, transcript)
 
-    variant_stop_codon_index = variant_protein.find("*")
-
     # genomic position to codon position
     aa_pos = int(cds_offset / 3)
 
-    # if contained stop codon, truncate sequence before it
-    if variant_stop_codon_index > -1:
-        variant_protein = variant_protein[:variant_stop_codon_index]
-
     if original_protein == variant_protein:
-        if aa_pos < len(original_protein):
+        original_start_codon = cds_seq[:3]
+        variant_start_codon = variant_cds_seq[:3]
+        if original_start_codon != variant_start_codon:
+            # mutation is silent on the amino acid sequence but
+            # uses a different start codon, which may cause the transcript
+            # to not be translated or translated in unexpected ways
+            return AlternateStartCodon(
+                variant,
+                transcript,
+                original_start_codon,
+                variant_start_codon)
+        elif aa_pos < len(original_protein):
             aa_ref = original_protein[aa_pos]
         elif aa_pos == len(original_protein):
             aa_ref = "*"
         elif aa_pos > len(original_protein):
-            logging.warn("How did we get aa_pos = %d when len(protein) = %d?",
-                aa_pos, len(original_protein))
+            if cds_seq[:len(original_protein) * 3 + 3].endswith("TGA"):
```

Comment:
Maybe add a quick comment giving the rationale (and maybe a reference) for this selenocysteine logic?